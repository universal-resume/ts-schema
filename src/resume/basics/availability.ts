import { NonEmptyTrimmedString } from "effect/Schema";

export const Availability = NonEmptyTrimmedString.annotations({
	description:
		"Date, delay, or any information about your availability to take a new job",
	examples: ["ASAP", "3 months", "September", "Q2"],
});
