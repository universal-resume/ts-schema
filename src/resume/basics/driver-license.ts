import { NonEmptyString } from "effect/Schema";

export const DriverLicense = NonEmptyString.annotations({
	description: "One of the driver licenses of the resume subject",
	examples: ["A", "B", "C"],
});
