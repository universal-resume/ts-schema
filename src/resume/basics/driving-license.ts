import { NonEmptyString } from "effect/Schema";

export const DrivingLicense = NonEmptyString.annotations({
	examples: ["A", "B", "C"],
});
