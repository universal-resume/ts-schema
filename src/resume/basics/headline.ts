import { maxLength, NonEmptyString } from "effect/Schema";

export const Headline = NonEmptyString.pipe(maxLength(50)).annotations({
	description: "The headline, title or position of resume subject",
	examples: ["Software Engineer"],
});
