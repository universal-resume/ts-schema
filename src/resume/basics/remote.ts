import { Boolean as BooleanSchema } from "effect/Schema";

export const Remote = BooleanSchema.annotations({
	description: "Whether the resume subject is willing to work remotely",
	examples: [true],
});
