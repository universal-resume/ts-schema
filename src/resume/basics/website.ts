import { Url } from "#value-object";

export const Website = Url.FromString.annotations({
	description: "The personal website, portfolio or blog of resume subject",
	examples: [Url.Brand.make(new URL("https://john-doe.me"))],
});
