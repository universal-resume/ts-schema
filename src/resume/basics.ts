import { Array as ArraySchema, optional, Struct } from "effect/Schema";
import {
	Contact,
	CountryCode,
	Day,
	Location,
	Profile,
	ProperNoun,
	Summary,
	Url,
} from "#value-object";
import { Availability } from "./basics/availability.js";
import { DrivingLicense } from "./basics/driving-license.js";
import { Headline } from "./basics/headline.js";
import { Remote } from "./basics/remote.js";
import { Website } from "./basics/website.js";

export namespace Basics {
	export const Schema = Struct({
		availability: optional(Availability),
		birth: optional(Day.FromString),
		drivingLicenses: optional(ArraySchema(DrivingLicense)),
		contact: optional(Contact.Schema),
		headline: Headline,
		picture: optional(Url.FromString),
		location: optional(Location.Schema),
		name: ProperNoun.FromString,
		nationalities: optional(ArraySchema(CountryCode.FromString)),
		profiles: optional(ArraySchema(Profile.Schema)),
		summary: optional(Summary.FromString),
		website: optional(Website),
		remote: optional(Remote),
	});

	export type Type = typeof Schema.Type;
}
