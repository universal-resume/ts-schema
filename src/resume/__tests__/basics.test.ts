import { Schema } from "effect";
import { expect, it } from "vitest";
import { Basics } from "../basics";

it("should parse this Basics", () => {
	const input: unknown = {
		name: "Bill Palmer",
		headline: "Vice President of IT Operations",
		summary:
			"Experienced IT leader with a strong background in IT operations, infrastructure management, and organizational transformation. Proven ability to lead cross-functional teams, modernize legacy systems, and drive digital transformation under pressure. Known for turning crisis into opportunity, and aligning IT with business goals.",
		birth: "1975-08-12",
		email: "bill.palmer@example.com",
		phone: "+1-555-123-4567",
		location: {
			address: "123 Enterprise Blvd",
			city: "Phoenix",
			region: "Arizona",
			postalCode: "85001",
			countryCode: "US",
		},
		nationalities: ["US"],
		driverLicenses: ["B"],
		picture: "https://example.com/images/bill-palmer.jpg",
		profiles: [
			{
				network: "LinkedIn",
				url: "https://www.linkedin.com/in/bill-palmer-it",
				username: "bill-palmer-it",
			},
		],
		website: "https://billpalmer.dev",
		remote: true,
	};
	const basics = Schema.decodeUnknownSync(Basics.Schema)(input);
	expect(basics).toMatchSnapshot();
});
