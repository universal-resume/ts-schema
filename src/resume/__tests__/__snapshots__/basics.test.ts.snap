// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should parse this Basics 1`] = `
{
  "birth": 1975-08-12T00:00:00.000Z,
  "headline": "Vice President of IT Operations",
  "location": {
    "address": "123 Enterprise Blvd",
    "city": "Phoenix",
    "countryCode": "US",
    "postalCode": "85001",
    "region": "Arizona",
  },
  "name": "Bill Palmer",
  "nationalities": [
    "US",
  ],
  "picture": "https://example.com/images/bill-palmer.jpg",
  "profiles": [
    {
      "network": "LinkedIn",
      "url": "https://www.linkedin.com/in/bill-palmer-it",
      "username": "bill-palmer-it",
    },
  ],
  "remote": true,
  "summary": "Experienced IT leader with a strong background in IT operations, infrastructure management, and organizational transformation. Proven ability to lead cross-functional teams, modernize legacy systems, and drive digital transformation under pressure. Known for turning crisis into opportunity, and aligning IT with business goals.",
  "website": "https://billpalmer.dev/",
}
`;
